File: backend/app/schemas/sla.py
from pydantic import BaseModel, Field
import uuid
from datetime import datetime
from app.models.ticket import TicketPriority
from app.models.ticket_sla import SLAStatus

class SLAPolicyBase(BaseModel):
    name: str = Field(..., max_length=100)
    priority: TicketPriority
    response_time_minutes: int = Field(..., gt=0)
    resolution_time_minutes: int = Field(..., gt=0)
    at_risk_threshold_percent: int = Field(..., ge=1, le=99)
    active: bool = True

class SLAPolicyCreate(SLAPolicyBase):
    pass

class SLAPolicyUpdate(BaseModel):
    name: str | None = Field(None, max_length=100)
    priority: TicketPriority | None = None
    response_time_minutes: int | None = Field(None, gt=0)
    resolution_time_minutes: int | None = Field(None, gt=0)
    at_risk_threshold_percent: int | None = Field(None, ge=1, le=99)
    active: bool | None = None

class SLAPolicyResponse(SLAPolicyBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class TicketSLAResponse(BaseModel):
    id: uuid.UUID
    ticket_id: uuid.UUID
    policy_id: uuid.UUID
    response_started_at: datetime
    response_due_at: datetime
    response_completed_at: datetime | None
    resolution_started_at: datetime
    resolution_due_at: datetime
    resolution_completed_at: datetime | None
    status: SLAStatus
    paused_at: datetime | None
    total_pause_seconds: int
    breached_at: datetime | None
    created_at: datetime
    updated_at: datetime
    policy: SLAPolicyResponse

    class Config:
        from_attributes = True

class SLAOverviewResponse(BaseModel):
    total_active_tickets: int
    on_track_count: int
    at_risk_count: int
    breached_count: int
    completed_count: int
    response_sla_compliance: float
    resolution_sla_compliance: float
    average_response_minutes: float
    average_resolution_minutes: float

File: frontend/src/services/sla.ts
import { apiClient } from './api';

export type SLAStatus = 'ON_TRACK' | 'AT_RISK' | 'BREACHED' | 'PAUSED' | 'COMPLETED';

export interface SLAPolicy {
  id: string;
  name: string;
  priority: string;
  response_time_minutes: number;
  resolution_time_minutes: number;
  at_risk_threshold_percent: number;
  active: boolean;
  created_at: string;
  updated_at: string;
}

export interface SLAPolicyCreate {
  name: string;
  priority: string;
  response_time_minutes: number;
  resolution_time_minutes: number;
  at_risk_threshold_percent: number;
  active?: boolean;
}

export interface SLAPolicyUpdate {
  name?: string;
  priority?: string;
  response_time_minutes?: number;
  resolution_time_minutes?: number;
  at_risk_threshold_percent?: number;
  active?: boolean;
}

export interface SLAOverviewResponse {
  total_active_tickets: number;
  on_track_count: number;
  at_risk_count: number;
  breached_count: number;
  completed_count: number;
  response_sla_compliance: number;
  resolution_sla_compliance: number;
  average_response_minutes: number;
  average_resolution_minutes: number;
}

export const slaService = {
  getPolicies: async () => {
    const response = await apiClient.get<SLAPolicy[]>('/sla/policies');
    return response.data;
  },

  createPolicy: async (data: SLAPolicyCreate) => {
    const response = await apiClient.post<SLAPolicy>('/sla/policies', data);
    return response.data;
  },

  updatePolicy: async (id: string, data: SLAPolicyUpdate) => {
    const response = await apiClient.patch<SLAPolicy>(`/sla/policies/${id}`, data);
    return response.data;
  },

  deactivatePolicy: async (id: string) => {
    await apiClient.delete(`/sla/policies/${id}`);
  },

  getOverview: async () => {
    const response = await apiClient.get<SLAOverviewResponse>('/sla/overview');
    return response.data;
  }
};
